vmalert: add `eval_offset` for group

Haleygo opened the pull request:
releated https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3409

Haleygo commented:
I restart the group here like proposed in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4634, otherwise the code will be too complicated when group eval_offset change[perform first evaluation after offset and reset ticker], and I don't think we want that. 
Then `interval` should be treated the same way.
please cc @hagen1778

hagen1778 commented on app/vmalert/config/config.go:
```diff
@@ -52,22 +54,43 @@ func (g *Group) UnmarshalYAML(unmarshal func(interface{}) error) error {
 
-// Validate check for internal Group or Rule configuration errors
+// SetDefaultFields sets some fields with default values if nil
+func (g *Group) SetDefaultFields(defaultGroupInterval time.Duration) {
```
I don't think we need to move this code to another method, especially public method. It gives not benefits from my perspective

Haleygo replied:
Fixed!

hagen1778 commented on app/vmalert/config/config.go:
```diff
@@ -208,24 +231,24 @@ type ValidateTplFn func(annotations map[string]string) error
 
 // Parse parses rule configs from given file patterns
-func Parse(pathPatterns []string, validateTplFn ValidateTplFn, validateExpressions bool) ([]Group, error) {
+func Parse(pathPatterns []string, validateTplFn ValidateTplFn, validateExpressions bool, defaultGroupInterval time.Duration) ([]Group, error) {
```
Would be nice to move the flag with default eval interval to config. But we should have done this from the very beginning. Now, evalInteval kinda belongs logically to main, as main decides when to evaluate something. So I think we shouldn't propagate `defaultGroupInterval` param to config. Let main.go to decide the default.

Haleygo replied:
Okay, brought `defaultGroupInterval` back to main.

hagen1778 commented on app/vmalert/group.go:
```diff
@@ -277,14 +281,30 @@ var skipRandSleepOnGroupStart bool
-	// Spread group rules evaluation over time in order to reduce load on VictoriaMetrics.
+	evalTS := time.Now()
+
+	// If group `eval_offset` is specified, will evaluate group
```
I don't think eval_offset should have effect on this delay.
Let's think about this in the following perspective:
1. There is a moment of time when rule is evaluated
2. And there is a moment of time with which this rule is evaluated.

This delay here affects p1 - it spreads evaluations over time to avoid many concurrent requests. 
But p2 is what actually eval_offset should have effect on. If I have a rule with `interval: 1h` and I want it to be evaluated on the 5th minute of the hour - I'm going to set `eval_offset: 5m`. But, this has nothing to do with p1. This rule can be evaluated on any moment of time on interval between 5th and 60th min of the current hour. I can evaluate it at 11:49 with timestamp 11:05, and it still will be correct.

So what I propose is to propagate eval_offect as one of https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0df506de54df5c24ef040fc2363ff402d5e80eb3/app/vmalert/datasource/vm.go#L33 new params. This new param will be checked here https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0df506de54df5c24ef040fc2363ff402d5e80eb3/app/vmalert/datasource/vm_prom_api.go#L167-L169  and if it is not nil - the current timestamp should be truncated by eval_offset.

It is also interesting that `datasource.queryTimeAlignment` kinda conflicts with eval_offset. I propose to resolve the conflict in the following way. At https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0df506de54df5c24ef040fc2363ff402d5e80eb3/app/vmalert/datasource/vm_prom_api.go#L167-L169  we'll do the following check:
1. If eval_offset is empty and `datasource.queryTimeAlignment=true` - we truncate timestamp with eval_interval
2.  If eval_offset is empty and `datasource.queryTimeAlignment=false` - we do nothing
3. If eval_offset is **not empty** and `datasource.queryTimeAlignment=true`- we truncate timestamp with eval_offset
4. If eval_offset is **not empty** and `datasource.queryTimeAlignment=false`- we truncate timestamp with eval_offset

wdyt @Haleygo ?

Haleygo replied:
>This rule can be evaluated on any moment of time on interval between 5th and 60th min of the current hour. I can evaluate it at 11:49 with timestamp 11:05, and it still will be correct.

Yeah, we can have rule evaluated at 11:49 with timestamp 11:05, but the result will only show up after 11:49 in that case, so it's still uncontrollable from my perspective. I use this `eval_offset` as a force offset for group evaluation in range of [0...group_interval] after aligned with interval.

>It is also interesting that datasource.queryTimeAlignment kinda conflicts with eval_offset. 
I mixed them before, then I divided their impacts, `eval_offset` only for evaluation time, `queryTimeAlignment` only for raw series timestamp of evaluation. 
If `eval_offset` is setting, the evaluation result is always consistent and predictable without effects from different vmalert or start time, so `queryTimeAlignment` is not needed. But I didn't see much influences here and left it as before.

Haleygo replied:
I think let `eval_offset` to control the p1 is somehow more clearly for users, like the original issue asked.
>If we could give it an evaluation-offset of 2m or something like that, that would make sure it evaluates and writes the result within the first 5m of the hour

People would want to see results at 11:02 when they set 2m offset instead of having unknown delay on evaluation.

hagen1778 replied:
> but the result will only show up after 11:49 in that case

just to clarify: yes, it will be queryable after 11:49, but it will have 11:05 timestamp.

Although, this 11:49 thing is what I'd like to fix afterwards. I think we should have a 5-10 min cap on the delay. For example, if we have 10 rules with interval 1h - all of them should start in first, let's say, 5min after vmalert start. I believe 5min interval is enough period to spread the load. So once we have this cap, the argument about delay becomes less relevant.

hagen1778 replied:
> People would want to see results at 11:02 when they set 2m offset instead of having unknown delay on evaluation.

I don't think people care that much about waiting for each result to appear. They care only about right timestamps and calculation on the chart.

Haleygo replied:
>All rules for a group with eval_offsset will start in the same moment of time

Actually rules in a group are evaluated at same time anyway. But mulitple groups with same offset will be crowded at same time if misconfigured.
But I get your points now, maybe my group example with 1h interval is too big, and took the deviation between 11:49 and 11:05 too serious.

hagen1778 replied:
> Actually rules in a group are evaluated at same time anyway. But mulitple groups with same offset will be crowded at same time if misconfigured.

yeah, my bad. Thanks for pointing that out

Haleygo replied:
>then we'll have the following consequences:

All rules for a group with eval_offsset will start in the same moment of time
There will be a conflict between the moment when group was evaluated and -datasource.queryTimeAlignment param
There will be a conflict between the moment when group was evaluated and -datasource.lookback param
Rules executed at timestamps close to the realtime are more likely to get a partial result, as data usually appears in the datasource with delay.

For now, this pr will address those problems:
1. All the group with same `eval_offset`  won’t be start at the same moment when vmalert restart, there will still be randSleep on each of group to spread the pressure like before.
2. Both `eval_offset` and `-datasource.queryTimeAlignment` have the power to influence the result of group evaluation.  `eval_offset`  will do it on group level, if `eval_offset`  is used, group’s evaluation results will always be at the exact offset in the range of [0...interval].
   `-datasource.queryTimeAlignment` will align the evaluation timestamp globally like before, but has no influence on group with `eval_offset`
3. `-datasource.lookback` won’t work when `eval_offset` is used. If users see partial results on group with `eval_offset`, it could be a rather big delay on the original data in datasource, it’s recommended to adding `-search.latencyOffset` on group’s params instead of using `lookback` , cause `lookback` will mess up the timestamp that result should have.
   FYI, this situation should be rarely, since group with `eval_offset` will sleep enough time to make sure the evaluation request happens after the offset ts. But if users do have a big delay on their original data, for now, they will have to set `search.latencyOffset`  for groups with `eval_offset` and `-datasource.lookback` for groups without `eval_offset`
4. see p3

hagen1778 commented on app/vmalert/datasource/vm_prom_api.go:
```diff
@@ -161,12 +161,26 @@ func (s *VMStorage) setPrometheusInstantReqParams(r *http.Request, query string,
-	if s.lookBack > 0 {
-		timestamp = timestamp.Add(-s.lookBack)
+
+	if s.evaluationInterval > 0 {
```
I think it is time to move this code to a function. This should simplify the code and make it more testable

Haleygo replied:
Good point, added it!

hagen1778 commented on app/vmalert/group.go:
```diff
@@ -163,6 +167,10 @@ func (g *Group) ID() uint64 {
 	hash.Write([]byte("\xff"))
 	hash.Write([]byte(g.Name))
 	hash.Write([]byte(g.Type.Get()))
+	hash.Write([]byte(g.Interval.String()))
+	if g.EvalOffset != nil {
+		hash.Write([]byte(g.EvalOffset.String()))
+	}
```
I don't think this is needed. Why do you think we should do this?

Haleygo replied:
See https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4693#issuecomment-1646561507.

hagen1778 commented on app/vmalert/group.go:
```diff
@@ -380,14 +409,12 @@ func (g *Group) start(ctx context.Context, nts func() []notifier.Notifier, rw *r
 
 			e.notifierHeaders = g.NotifierHeaders
 
-			if g.Interval != ng.Interval {
```
Why this code was deleted?

hagen1778 commented on docs/CHANGELOG.md:
```diff
@@ -32,6 +32,8 @@ The following `tip` changes can be tested by building VictoriaMetrics components
 * FEATURE: [vmui](https://docs.victoriametrics.com/#vmui): improve accessibility score to 100 according to [Google's Lighthouse](https://developer.chrome.com/docs/lighthouse/accessibility/?utm_source=lighthouse&utm_medium=devtools) tests.
 * FEATURE: [vmui](https://docs.victoriametrics.com/#vmui): organize `min`, `max`, `median` values on the chart legend and tooltips for better visibility.
+* FEATURE: [vmalert](https://docs.victoriametrics.com/vmalert.html): add `eval_offset` attribute for rule group. If specified, group will be evaluated at the nearest offset time point aligned with interval, using random offset if not. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3409) for details.
+* FEATURE: [vmalert](https://docs.victoriametrics.com/vmalert.html): add `eval_offset` attribute for rule group. If specified, group will be evaluated at the exact offset in the range of [0...interval]. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3409) for details.
```
an extra feature)

Haleygo replied:
Sorry*2, fixed!

hagen1778 commented on docs/vmalert.md:
```diff
@@ -123,6 +123,11 @@ name: <string>
 # How often rules in the group are evaluated.
 [ interval: <duration> | default = -evaluationInterval flag ]
 
+# Group will be evaluated at the nearest offset time point aligned with interval.
```
doc isn't synced

Haleygo replied:
Sorry, fixed!

Haleygo commented on app/vmalert/config/config.go:
```diff
@@ -63,11 +64,24 @@ func (g *Group) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	}
+	if g.Interval.Duration() < 0 || g.EvalOffset.Duration() < 0 {
+		return fmt.Errorf("neither interval nor eval_offset should be less than 0")
+	}
```
Fixed!

Haleygo commented on app/vmalert/alerting_test.go:
```diff
@@ -872,7 +882,6 @@ func TestAlertingRule_Template(t *testing.T) {
 				gotAlert := tc.rule.alerts[hash]
 				if gotAlert == nil {
 					t.Fatalf("alert %d is missing; labels: %v; annotations: %v", hash, expAlert.Labels, expAlert.Annotations)
-					break
```
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/39937150/84a9c59c-3e67-407d-ba9f-6707a3da3cb6)
Weird, golangci-lint will warning if I got this line.

hagen1778 commented:
Hey @Haleygo! I've added a few changes to this PR. But what concerns me is the test case here https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4693/commits/a302c0d02175837a368a2a2a3e88d67ee9bab883#diff-c939f4dddd527e90292421832544f1a90b16a686807e00168377dd205dcceff9.
Could you please verify that eval_offset delay working as expected?

Haleygo commented:
> Hey @Haleygo! I've added a few changes to this PR. But what concerns me is the test case here [a302c0d#diff-c939f4dddd527e90292421832544f1a90b16a686807e00168377dd205dcceff9](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/a302c0d02175837a368a2a2a3e88d67ee9bab883#diff-c939f4dddd527e90292421832544f1a90b16a686807e00168377dd205dcceff9). Could you please verify that eval_offset delay working as expected?

It's works as expected, there are too much variable on delay calculation, I introduced a fixed origin randSleep to show the situation where offset will work, please check)

hagen1778 commented:
LGTM overall. But would you mind checking changes here https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4693/commits/e9da8fece9370fb81d450074de0720fc950e398e ?

Haleygo commented:
> LGTM overall. But would you mind checking changes here [e9da8fe](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/e9da8fece9370fb81d450074de0720fc950e398e) ?

Thanks! LGTM! 
Found some minor typo, and I don't think `time.DateTime` is valid in builtin `time` package, do you use some custom package here?
```
		at, err := time.Parse(time.DateTime, atS)
		if err != nil {
			t.Fatal(err)
		}
		expTS, err := time.Parse(time.DateTime, expS)
```

hagen1778 reviewed (approved):
LGTM
